samtranslator/swagger/swagger.py

```diff
@@ -799,26 +802,193 @@ def add_resource_policy(self, resource_policy):
 
-                self.resource_policy['Statement'] = statement
+        if ip_deny_list is not None:
+            resource_list = self._get_method_path_uri_list(path, api_id, stage)
+            self._add_ip_resource_policy_for_method(ip_deny_list, "IpAddress", resource_list)
 
+        if source_vpc_allow_list is not None:
+            resource_list = self._get_method_path_uri_list(path, api_id, stage)
+            for endpoint in source_vpc_allow_list:
+                self._add_vpc_resource_policy_for_method(endpoint, "StringNotEquals", resource_list)
+
+        if source_vpc_deny_list is not None:
+            resource_list = self._get_method_path_uri_list(path, api_id, stage)
+            for endpoint in source_vpc_deny_list:
+                self._add_vpc_resource_policy_for_method(endpoint, "StringEquals", resource_list)
+
+        if custom_statements is not None:
+            self._add_custom_statement(custom_statements)
+
+        if self._doc.get(self._X_APIGW_POLICY) is None:
             self._doc[self._X_APIGW_POLICY] = self.resource_policy
+        else:
+            apigw_policy = self._doc[self._X_APIGW_POLICY]
+            apigw_policy['Version'] = resource_policy['Version']
+            apigw_policy['Statement'].extend(resource_policy.get('Statement'))
+            self._doc[self._X_APIGW_POLICY] = apigw_policy
+
+    def _add_iam_resource_policy_for_method(self, policy_list, effect, resource_list):
+        if not policy_list:
+            return
+
+        if effect not in ["Allow", "Deny"]:
+            return
+
+        if not isinstance(policy_list, list):
+            policy_list = [policy_list]
+
+        self.resource_policy['Version'] = '2012-10-17'
+        policy_statement = {}
+        policy_statement['Effect'] = effect
+        policy_statement['Action'] = "execute-api:Invoke"
+        policy_statement['Resource'] = resource_list
+        policy_statement['Principal'] = {"AWS": policy_list}
+
+        if self.resource_policy.get('Statement') is None:
+            self.resource_policy['Statement'] = policy_statement
+        else:
+            statement = self.resource_policy['Statement']
+            if not isinstance(statement, list):
+                statement = [statement]
+            statement.extend(policy_statement)
+            self.resource_policy['Statement'] = statement
+
+    def _get_method_path_uri_list(self, path, api_id, stage):
+        # It turns out that APIGW doesn't like trailing slashes in paths (#665)
+        # and removes as a part of their behaviour, but this isn't documented.
+        # The regex removes the tailing slash to ensure the permission works as intended
+        # At this point, value of Swagger path should be a dictionary with method names being the keys
+        methods = list(self.get_path(path).keys())
+        uri_list = []
+        path = re.sub(r'{([a-zA-Z0-9._-]+|proxy\+)}', '*', path)
```

Comment:
It looks like this regex is copied from elsewhere, could we keep the regex in one place as a variable and use that variable in multiple places?